### PR TITLE
Add bm65hsrgb/rev2

### PR DESCRIPTION
**Pull request by peepeetee**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* #15081
* #14964

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kprepublic/bm65rgb/rev2/rev2.h [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (drashna) [inline keyboards/kprepublic/bm65rgb/rev2/rev2.h:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by rotaryking** [comment]
After flashing on my BM65V2, it works successfully

**Review by drashna — APPROVED** [review]

File: keyboards/kprepublic/bm65hsrgb/rev2/rules.mk [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/kprepublic/bm65hsrgb/rev2/rules.mk:16]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/kprepublic/bm65hsrgb/rev2/rules.mk [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+LTO_ENABLE = yes
+
+# RGB Matrix is required to support per-key LEDs connected to IS31FL3741.
+RGB_MATRIX_ENABLE = yes
+
+# The simpler RGB Lighting code can be enabled to control just the underglow
+# LEDs from it; leave it disabled to make the RGB Matrix code control all LEDs,
+# including underglow.
+RGBLIGHT_ENABLE = no
+
```
Comment (fauxpark) [inline keyboards/kprepublic/bm65hsrgb/rev2/rules.mk:32]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/kprepublic/bm65hsrgb/rev2/rules.mk (hunk 3 above)
Comment (peepeetee) [reply keyboards/kprepublic/bm65hsrgb/rev2/rules.mk:32]: Why?

**Comment by rotaryking** [comment]
image
error within rev2.c

**Comment by peepeetee** [comment]
@rotaryking compiles fine after my latest commit, 
image
image

File: keyboards/kprepublic/bm65hsrgb/rev2/keymaps/peepeetee/config.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+// #define RGB_DISABLE_AFTER_TIMEOUT 0   // number of ticks to wait until disabling effects
+// #define RGB_DISABLE_WHEN_USB_SUSPENDED false // turn off effects when suspended
+// #define RGB_MATRIX_LED_PROCESS_LIMIT (DRIVER_LED_TOTAL + 4) / 5 // limits the number of LEDs to process in an animation per task run (increases keyboard responsiveness)
+// #define RGB_MATRIX_LED_FLUSH_LIMIT 16 // limits in milliseconds how frequently an animation will update the LEDs. 16 (16ms) is equivalent to limiting to 60fps (increases keyboard responsiveness)
+
+
+// This list in in the correct mode order. Next mode is the following line, previous mode is previous line. Loops around.
+// #define DISABLE_RGB_MATRIX_SOLID_COLOR                   // Static single hue, no speed support
+// #define DISABLE_RGB_MATRIX_ALPHAS_MODS                   // Static dual hue, speed is hue for secondary hue
+#define DISABLE_RGB_MATRIX_GRADIENT_UP_DOWN              // Static gradient top to bottom, speed controls how much gradient changes
```
Comment (drashna) [inline keyboards/kprepublic/bm65hsrgb/rev2/keymaps/peepeetee/config.h:58]: These need to be updated.

File: keyboards/kprepublic/bm65hsrgb/rev2/keymaps/peepeetee/config.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+#pragma once
+
+
+
+// #define IS_COMMAND() (get_mods() == MOD_MASK_SHIFT)      // Key combination that allows the use of magic commands (useful for debugging)
+
+// #define TERMINAL_HELP
+
+#define FORCE_NKRO                  // NKRO by default requires to be turned on, this forces it on during keyboard startup regardless of EEPROM setting. NKRO can still be turned off but will be turned on again if the keyboard reboots.
+#define USB_POLLING_INTERVAL_MS 1
```
Comment (drashna) [inline keyboards/kprepublic/bm65hsrgb/rev2/keymaps/peepeetee/config.h:25]: May not be worth having, as this is the default on develop, and due to be merged soon

Comment (drashna) [inline keyboards/kprepublic/bm65hsrgb/rev2/keymaps/peepeetee/config.h:25]: May not be worth having, as this is the default on develop, and due to be merged soon

File: keyboards/kprepublic/bm65hsrgb/rev2/rules.mk [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+LTO_ENABLE = yes
+
+LAYOUTS = 65_ansi_blocker
+
+LTO_ENABLE = yes
+
+# RGB Matrix is required to support per-key LEDs connected to IS31FL3741.
+RGB_MATRIX_ENABLE = yes
```
Comment (drashna) [inline keyboards/kprepublic/bm65hsrgb/rev2/rules.mk:26]: Would probably be a good idea to add: 
```suggestion
RGB_MATRIX_ENABLE = yes
RGBLIGHT_SUPPORTED = no
```

Comment (drashna) [inline keyboards/kprepublic/bm65hsrgb/rev2/rules.mk:26]: Would probably be a good idea to add: 
```suggestion
RGB_MATRIX_ENABLE = yes
RGBLIGHT_SUPPORTED = no
```

File: keyboards/kprepublic/bm65hsrgb/rev2/rules.mk [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+NKRO_ENABLE = yes           # Enable N-Key Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+LTO_ENABLE = yes
+
+LAYOUTS = 65_ansi_blocker
+
+LTO_ENABLE = yes
+
```
Comment (fauxpark) [inline keyboards/kprepublic/bm65hsrgb/rev2/rules.mk:24]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by 4itor** [comment]
Sorry about the noob question, is it possible to compile QMK for the board (rev2) before this pull-request gets finished?

Thanks in advance.

PD: Wanting to play with tap-dance and combos so bad :)

**Comment by 4itor** [comment]
Please, what is missing for the pull request to succeed?
Can we help?

**Comment by peepeetee** [comment]
Rebased just now, should have resolved the conflicts

**Comment by drashna** [comment]
> ☒ kprepublic/bm65hsrgb/rev2: NO_ACTION_MACRO in config.h is no longer a valid option
> ☒ kprepublic/bm65hsrgb/rev2: NO_ACTION_FUNCTION in config.h is no longer a valid option

**Comment by CerebroJD** [comment]
Is there an opportunity to get these changes merged into qmk so these v2 boards are supported?

**Comment by chigzy** [comment]
What’s remaining for this?
Anywhere to help?

**Comment by peepeetee** [comment]
> What’s remaining for this? Anywhere to help?

just had to poke me, and you did

**Comment by noroadsleft** [comment]
This PR needs to be rebased upon a current `qmk:master` – there are a lot of merge conflicts, which makes this unable to be merged.

**Comment by peepeetee** [comment]
> This PR needs to be rebased upon a current `qmk:master` – there are a lot of merge conflicts, which makes this unable to be merged.

I need to undo [b67c072](https://github.com/qmk/qmk_firmware/pull/15060/commits/b67c072682f9b6bb1a61492f71e91169d72460a8) and rebase on master instead. how would I do that?

File: keyboards/kprepublic/bm65hsrgb/rev2/config.h [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+ * ROWS: AVR pins used for rows, top to bottom
+ * DIODE_DIRECTION: COL2ROW = COL = Anode (+), ROW = Cathode (-, marked on diode)
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ *
+ */
+
+// #define UNUSED_PINS
+
+/* COL2ROW, ROW2COL */
+//#define DIODE_DIRECTION COL2ROW
```
Comment (drashna) [inline keyboards/kprepublic/bm65hsrgb/rev2/config.h:50]: This should be removed, not just commented out.

File: keyboards/kprepublic/bm65hsrgb/rev2/config.h [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+// The pin connected to the data pin of the LEDs
+#define RGB_DI_PIN E2
+// The number of LEDs connected
+#define DRIVER_LED_TOTAL 73
+#ifdef RGB_DI_PIN
+    #define RGBLED_NUM 6
```
Comment (drashna) [inline keyboards/kprepublic/bm65hsrgb/rev2/config.h:57]: These may need to be updated.

**Comment by AstralCookie** [comment]
Any updates on this? Just got myself a BM65V2 and I want to change the backlight's LED's color

**Pull request closed without merging**